Source/Core/InputCommon/ControllerInterface/Device.cpp

```diff
@@ -146,15 +146,43 @@ bool DeviceQualifier::operator==(const DeviceQualifier& devq) const
 
 Device* DeviceContainer::FindDevice(const DeviceQualifier& devq) const
 {
-  for (Device* d : m_devices)
+  std::lock_guard<std::mutex> lk(m_devices_mutex);
+  for (const auto& d : m_devices)
   {
-    if (devq == d)
-      return d;
+    if (devq == d.get())
+      return d.get();
```

Comment:
This method is very unsafe with the current threading model. If we want other objects to be able to reference Device objects owned by this class, we should use shared pointers.